shared/src/server/proxy_protocol.rs

```diff
@@ -0,0 +1,257 @@
+const MIN_PROXY_PROTOCOL_HEADER_LEN: usize = 16;
+
+pub struct AcceptedConn<S: AsyncRead + AsyncWrite> {
+    inner: S,
+    proxy_protocol: Option<PPHeader<'static>>,
+    preread_buffer: Option<Vec<u8>>,
+}
+
+impl<S: AsyncRead + AsyncWrite + Sync> AcceptedConn<S> {
+    fn new(
+        inner: S,
+        proxy_protocol: Option<PPHeader<'static>>,
+        opt_preread_buffer: Option<Vec<u8>>,
+    ) -> Self {
+        // if the preread buffer is empty, we don't need to store it
+        let preread_buffer = match opt_preread_buffer.as_deref() {
+            Some(buf) if buf.len() == 0 => None,
+            Some(_) => opt_preread_buffer,
+            None => None,
+        };
+        Self {
+            inner,
+            proxy_protocol,
+            preread_buffer,
+        }
+    }
+}
+
+impl<S: AsyncRead + AsyncWrite + Unpin + Sync> AsyncRead for AcceptedConn<S> {
+    fn poll_read(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<std::io::Result<()>> {
+        let this = self.get_mut();
+        if let Some(preread_buf) = this.preread_buffer.as_mut() {
+            let len = std::cmp::min(preread_buf.len(), buf.remaining());
+            buf.put_slice(&preread_buf[..len]);
+            preread_buf.drain(..len);
+            if preread_buf.is_empty() {
+                this.preread_buffer = None;
+            }
+            return std::task::Poll::Ready(Ok(()));
+        }
+        std::pin::Pin::new(&mut this.inner).poll_read(cx, buf)
+    }
+}
+
+impl<S: AsyncRead + AsyncWrite + Unpin + Sync> AsyncWrite for AcceptedConn<S> {
+    fn poll_write(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<std::result::Result<usize, std::io::Error>> {
+        let pinned_inner = std::pin::Pin::new(&mut self.get_mut().inner);
+        pinned_inner.poll_write(cx, buf)
+    }
+
+    fn poll_flush(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<std::result::Result<(), std::io::Error>> {
+        let pinned_inner = std::pin::Pin::new(&mut self.get_mut().inner);
+        pinned_inner.poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<std::result::Result<(), std::io::Error>> {
+        let pinned_inner = std::pin::Pin::new(&mut self.get_mut().inner);
+        pinned_inner.poll_shutdown(cx)
+    }
+}
+
+pub async fn try_parse_proxy_protocol<S: AsyncRead + AsyncWrite + Unpin + Sync>(
+    mut incoming_conn: S,
+) -> ServerResult<AcceptedConn<S>> {
+    let mut buf: Vec<u8> = Vec::with_capacity(MIN_PROXY_PROTOCOL_HEADER_LEN);
+    let mut total_read_bytes = 0;
+    loop {
+        let read_len = incoming_conn.read_buf(&mut buf).await?;
+        if read_len == 0 {
+            println!("empty read");
```

Comment:
Looks like there's a few debug logs left in?

Reply:
removed!